File: apps/arc_data/lib/arc/data/packet.ex
defmodule Arc.Data.Packet do
  @moduledoc """
  Arc wire packet format.

      [4 bytes]   header_length (big-endian uint32)
      [N bytes]   header (JSON): {src, dst, sid, seq, ts, ph}
      [64 bytes]  Ed25519 signature over header bytes
      [N bytes]   payload: nonce (12 bytes) <> ChaCha20-Poly1305 ciphertext

  Fields:
    src  base64(Ed25519 public key of sender)
    dst  base64(Ed25519 public key of recipient)
    sid  base64(session ID, 16 random bytes established at session setup)
    seq  monotonic sequence number per session (replay prevention)
    ts   unix milliseconds (freshness check)
    ph   base64(SHA-256(payload)) — binds payload to signed header

  The Ed25519 signature covers the raw header bytes, binding all header
  fields to the sender's keypair. A valid signature proves the sender
  holds the private key corresponding to src.

  ph binds the ciphertext to the signed header, preventing payload
  substitution attacks.

  Note: ph uses SHA-256 pending a Blake3 dependency (same security
  properties for this use case — wire compatibility will be maintained
  when upgraded).
  """

  alias Arc.Identity

  @signature_bytes 64
  @nonce_bytes 12

  @doc """
  Encode and sign a packet.
  """
  @spec encode(Identity.t(), binary(), binary(), non_neg_integer(), binary(), binary(), keyword()) ::
          binary()
  def encode(%Identity{} = src_identity, dst_pk, session_id, seq, nonce, ciphertext, opts \\ []) do
    payload = nonce <> ciphertext
    ts = Keyword.get(opts, :ts, System.system_time(:millisecond))

    header_map = %{
      "src" => Base.encode64(src_identity.public_key),
      "dst" => Base.encode64(dst_pk),
      "sid" => Base.encode64(session_id),
      "seq" => seq,
      "ts" => ts,
      "ph" => Base.encode64(:crypto.hash(:sha256, payload))
    }

    header_bytes = IO.iodata_to_binary(:json.encode(header_map))
    header_len = byte_size(header_bytes)
    signature = Identity.sign(src_identity, header_bytes)

    <<header_len::32-big>> <> header_bytes <> signature <> payload
  end

  @doc """
  Decode and verify a packet. Returns the decoded fields including
  separated nonce and ciphertext.

  Verifies:
    1. Ed25519 signature over header
    2. Payload hash matches ph in header
  """
  @spec decode(binary()) ::
          {:ok,
           %{
             src: binary(),
             dst: binary(),
             session_id: binary(),
             seq: non_neg_integer(),
             ts: non_neg_integer(),
             nonce: binary(),
             ciphertext: binary()
           }}
          | {:error, :malformed_packet | :invalid_signature | :payload_hash_mismatch}
  def decode(<<header_len::32-big, rest::binary>>) do
    case rest do
      <<header_bytes::binary-size(header_len), sig::binary-size(@signature_bytes),
        payload::binary>> ->
        verify_and_decode(header_bytes, sig, payload)

      _ ->
        {:error, :malformed_packet}
    end
  end

  def decode(_), do: {:error, :malformed_packet}

  defp verify_and_decode(header_bytes, sig, payload) do
    with {:ok, header} <- safe_decode_json(header_bytes),
         {:ok, src_pk} <- safe_decode64(header["src"]),
         {:sig, true} <- {:sig, Identity.verify(src_pk, header_bytes, sig)},
         expected_ph = Base.encode64(:crypto.hash(:sha256, payload)),
         {:ph, true} <- {:ph, expected_ph == header["ph"]},
         {:ok, dst_pk} <- safe_decode64(header["dst"]),
         {:ok, session_id} <- safe_decode64(header["sid"]),
         {:seq, true} <- {:seq, is_integer(header["seq"]) and header["seq"] >= 0},
         {:ts, true} <- {:ts, is_integer(header["ts"])},
         <<nonce::binary-size(@nonce_bytes), ciphertext::binary>> <- payload do
      {:ok,
       %{
         src: src_pk,
         dst: dst_pk,
         session_id: session_id,
         seq: header["seq"],
         ts: header["ts"],
         nonce: nonce,
         ciphertext: ciphertext
       }}
    else
      {:sig, false} -> {:error, :invalid_signature}
      {:ph, false} -> {:error, :payload_hash_mismatch}
      {:seq, false} -> {:error, :malformed_packet}
      {:ts, false} -> {:error, :malformed_packet}
      _ -> {:error, :malformed_packet}
    end
  end

  defp safe_decode_json(bytes) do
    try do
      {:ok, :json.decode(bytes)}
    rescue
      _ -> {:error, :malformed_packet}
    end
  end

  defp safe_decode64(nil), do: {:error, :malformed_packet}

  defp safe_decode64(str) do
    case Base.decode64(str) do
      {:ok, bytes} -> {:ok, bytes}
      :error -> {:error, :malformed_packet}
    end
  end
end

File: go/packet/packet.go
// Package packet holds the ARC wire packet.
//
//	[4 bytes]   header length, big endian uint32
//	[N bytes]   header, JSON: {src, dst, sid, seq, ts, ph, ek?}
//	[64 bytes]  Ed25519 signature over the header bytes
//	[N bytes]   payload: nonce, 12 bytes, then the ChaCha20-Poly1305 ciphertext
//
// Fields:
//
//	src  base64 of the Ed25519 public key of the sender
//	dst  base64 of the Ed25519 public key of the recipient
//	sid  base64 of the session id, 16 bytes from the session setup
//	seq  the sequence number of the session, which guards against replay
//	ts   unix milliseconds, which gives freshness
//	ph   base64 of SHA-256 of the payload, which binds the payload to the header
//	ek   base64 of the ephemeral X25519 public key of the initiator. Every
//	     session version 2 packet carries it. A version 1 packet does not.
//
// The signature covers the header bytes. A valid signature proves that the
// sender holds the secret key of src. The ph field stops an attacker from
// replacing the payload under a signed header.
//
// The hash is SHA-256 until a BLAKE3 change to the wire format.
package packet

import (
	"bytes"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"errors"
	"strconv"
	"time"

	"github.com/gezibash/arc/go/identity"
)

const (
	lengthBytes    = 4
	signatureBytes = 64
	// NonceBytes is the length of the nonce at the front of the payload.
	NonceBytes = 12
	// MaxHeaderBytes caps the header of a packet from an unknown peer.
	MaxHeaderBytes = 64 * 1024
)

// Errors of this package. A peer that is not trusted sends the bytes, so every
// failure returns one of these and never a panic.
var (
	ErrMalformed           = errors.New("packet: malformed packet")
	ErrInvalidSignature    = errors.New("packet: invalid signature")
	ErrPayloadHashMismatch = errors.New("packet: payload hash mismatch")
)

// Packet holds the fields of one decoded packet.
type Packet struct {
	Src             []byte
	Dst             []byte
	SessionID       []byte
	Seq             uint64
	Timestamp       int64
	EphemeralPublic []byte
	Nonce           []byte
	Ciphertext      []byte
}

// Option changes one field of an encoded packet.
type Option func(*options)

type options struct {
	timestamp       int64
	ephemeralPublic []byte
}

// WithTimestamp sets the timestamp in unix milliseconds. Without it the packet
// carries the time of the call.
func WithTimestamp(unixMilli int64) Option {
	return func(o *options) { o.timestamp = unixMilli }
}

// WithEphemeralKey adds the ephemeral X25519 public key of the session. Every
// session version 2 packet carries it.
func WithEphemeralKey(key []byte) Option {
	return func(o *options) { o.ephemeralPublic = key }
}

type header struct {
	Src string `json:"src"`
	Dst string `json:"dst"`
	SID string `json:"sid"`
	Seq uint64 `json:"seq"`
	TS  int64  `json:"ts"`
	PH  string `json:"ph"`
	EK  string `json:"ek,omitempty"`
}

// Encode builds one signed packet.
func Encode(me *identity.Identity, dst, sessionID []byte, seq uint64, nonce, ciphertext []byte, opts ...Option) ([]byte, error) {
	config := options{timestamp: time.Now().UnixMilli()}
	for _, opt := range opts {
		opt(&config)
	}

	payload := make([]byte, 0, len(nonce)+len(ciphertext))
	payload = append(payload, nonce...)
	payload = append(payload, ciphertext...)
	sum := sha256.Sum256(payload)

	head := header{
		Src: encode64(me.PublicKey),
		Dst: encode64(dst),
		SID: encode64(sessionID),
		Seq: seq,
		TS:  config.timestamp,
		PH:  encode64(sum[:]),
	}
	if len(config.ephemeralPublic) == identity.SeedBytes {
		head.EK = encode64(config.ephemeralPublic)
	}

	headerBytes, err := json.Marshal(head)
	if err != nil {
		return nil, err
	}

	out := make([]byte, lengthBytes, lengthBytes+len(headerBytes)+signatureBytes+len(payload))
	binary.BigEndian.PutUint32(out, uint32(len(headerBytes)))
	out = append(out, headerBytes...)
	out = append(out, me.Sign(headerBytes)...)
	return append(out, payload...), nil
}

// Decode reads one packet, checks the signature over the header, and checks
// the payload against the hash in the header.
func Decode(raw []byte) (*Packet, error) {
	if len(raw) < lengthBytes {
		return nil, ErrMalformed
	}

	headerLen := int(binary.BigEndian.Uint32(raw[:lengthBytes]))
	if headerLen > MaxHeaderBytes || len(raw) < lengthBytes+headerLen+signatureBytes {
		return nil, ErrMalformed
	}

	headerBytes := raw[lengthBytes : lengthBytes+headerLen]
	signature := raw[lengthBytes+headerLen : lengthBytes+headerLen+signatureBytes]
	payload := raw[lengthBytes+headerLen+signatureBytes:]

	head, err := decodeHeader(headerBytes)
	if err != nil {
		return nil, err
	}

	src, err := decode64(head.Src, identity.SeedBytes)
	if err != nil {
		return nil, err
	}
	if !identity.Verify(src, headerBytes, signature) {
		return nil, ErrInvalidSignature
	}

	sum := sha256.Sum256(payload)
	if head.PH != encode64(sum[:]) {
		return nil, ErrPayloadHashMismatch
	}

	dst, err := decode64(head.Dst, 0)
	if err != nil {
		return nil, err
	}

	sessionID, err := decode64(head.SID, 0)
	if err != nil {
		return nil, err
	}

	var ephemeral []byte
	if head.EK != "" {
		if ephemeral, err = decode64(head.EK, identity.SeedBytes); err != nil {
			return nil, err
		}
	}

	if len(payload) < NonceBytes {
		return nil, ErrMalformed
	}

	return &Packet{
		Src:             src,
		Dst:             dst,
		SessionID:       sessionID,
		Seq:             head.Seq,
		Timestamp:       head.TS,
		EphemeralPublic: ephemeral,
		Nonce:           payload[:NonceBytes],
		Ciphertext:      payload[NonceBytes:],
	}, nil
}

// decodeHeader refuses anything that the Elixir implementation refuses: a
// header that is not an object, a sequence number that is not a whole number
// at or above zero, and a timestamp that is not a whole number.
func decodeHeader(raw []byte) (*header, error) {
	var fields map[string]json.RawMessage
	decoder := json.NewDecoder(bytes.NewReader(raw))
	if err := decoder.Decode(&fields); err != nil {
		return nil, ErrMalformed
	}
	if decoder.More() {
		return nil, ErrMalformed
	}

	head := &header{}
	var err error
	if head.Seq, err = wholeNumber(fields["seq"]); err != nil {
		return nil, err
	}
	if head.TS, err = signedNumber(fields["ts"]); err != nil {
		return nil, err
	}

	for value, field := range map[*string]string{
		&head.Src: "src", &head.Dst: "dst", &head.SID: "sid", &head.PH: "ph",
	} {
		if err := json.Unmarshal(fields[field], value); err != nil {
			return nil, ErrMalformed
		}
	}
	if raw, ok := fields["ek"]; ok {
		if err := json.Unmarshal(raw, &head.EK); err != nil {
			return nil, ErrMalformed
		}
	}
	return head, nil
}

func wholeNumber(raw json.RawMessage) (uint64, error) {
	value, err := strconv.ParseUint(string(raw), 10, 64)
	if err != nil {
		return 0, ErrMalformed
	}
	return value, nil
}

// signedNumber accepts a negative timestamp, because the Elixir implementation
// accepts one. The freshness check refuses it later.
func signedNumber(raw json.RawMessage) (int64, error) {
	value, err := strconv.ParseInt(string(raw), 10, 64)
	if err != nil {
		return 0, ErrMalformed
	}
	return value, nil
}

func encode64(value []byte) string {
	return base64.StdEncoding.EncodeToString(value)
}

// decode64 decodes one base64 field. A want of zero accepts any length.
func decode64(value string, want int) ([]byte, error) {
	out, err := base64.StdEncoding.DecodeString(value)
	if err != nil || (want > 0 && len(out) != want) {
		return nil, ErrMalformed
	}
	return out, nil
}
